source/geod24/Serialization.d

```diff
@@ -0,0 +1,259 @@
+/*******************************************************************************
+
+    Serialization-related utilities
+
+    When passing instances accross threads, LocalRest often cannot use the
+    provided instance itself, as it would require it to be either `shared`,
+    a value type, or `immutable`.
+    The term instance here is used to mean an instance of a type,
+    that is either passed as parameter, and used as a return value.
+
+    Instead, we support serializing the parameter to one of three types:
+    `immutable(void)[]`, `immutable(ubyte)[]`, or `string`.
+    The first two are regular serialized target, and `string` can be used
+    for JSON, XML, etc.
+
+    A serializer can be a `template` or an aggregate, or even a function
+    returning an aggregate. No limitation is put on the type itself,
+    only on what operations should be supported.
+
+    For a type `T` to be a valid serializer, it should support
+    the following actions:
+    - `T.serialize()` should compile when passed on parameter.
+    - This one parameter can be any type that is either a parameter or a
+      return value of `API`'s methods. This is not checked explicitly,
+      but will lead to compilation error in LocalRest if not followed.
+    - `T.serialize(param)` should return an one of `string`,
+      `immutable(void)[]`, `immutable(ubyte)[]`
+    - `T.deserialize!QT` is a template function that returns an
+      instance of type `QT` (`QT` is potentially qualified, e.g. `const`).
+    - `T.deserialize!!T()` should accept one runtime non-default
+       parameter, of the type returned by `T.serialize`.
+
+    The default implementation is the `VibeJSONSerializer`.
+    Other utilities in this module are used to facilitate development
+    and diagnostics.
+
+    Author:         Mathias 'Geod24' Lang
+    License:        MIT (See LICENSE.txt)
+    Copyright:      Copyright (c) 2020 Mathias Lang. All rights reserved.
+
+*******************************************************************************/
+
+module geod24.Serialization;
+
+import std.traits;
+
+/*******************************************************************************
+
+    Serialize arbitrary data to / from JSON
+
+    This is the default serializer used by LocalRest, when none is provided.
+    It is a template so that LocalRest does not import Vibe.d if not used.
```

Comment:
Won't it be imported regardless when `-unittest` is passed in the client code? Because the unittest block below will instantiate this template.

Reply:
Unittests in non-root modules should not be parsed

Reply:
hmm.. this is when using dub only, right?

Reply:
No it's a DMD behavior, see discussions here: https://github.com/dlang/dmd/pull/10937

Reply:
Ah interesting. I know Walter implemented the "don't parse unittest blocks when -unittest is not passed", but I didn't know about this other feature.